Core/Instances: implemented handling for SMSG_BOSS_KILL_CREDIT

Ovahlord opened the pull request:
**Changes proposed**:
- Adds packet structure for the opcode and implements a instancescript function to call it via internal script or directly within a boss script that can access the instancescript.

**Target branch(es)**: 6x

**Tests performed**: Tested ingame, works as intended

The opcode connection type has not been changed to instance because the opcode can also be sent to players within the open world for example by killing Supreme Lord Kazzak.
The packet structure itself was not in the client function name list that I got via gist so I used the most common names of that list

The following screenshot has been taken by using my IronDocks branch at https://github.com/Ovahlord/TrinityCore/tree/IronDocks

![wowscrnshot_070516_121231](https://cloud.githubusercontent.com/assets/18347559/16616006/a16fa4fe-437a-11e6-838c-ca9717b39ff4.jpg)

Commit:
Core/Instances: implemented handling for SMSG_BOSS_KILL_CREDIT *the packet header names has not been found within the gist file so I used the names that made most sense *dungeon encounter id's can be found in the DungeonEncounter.dbc file

MitchesD commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -236,6 +236,15 @@ namespace WorldPackets
+        class InstanceBossKillCredit final : public ServerPacket
+        {
+        public:
+            InstanceBossKillCredit() : ServerPacket(SMSG_BOSS_KILL_CREDIT, 0) { }
```
just a small thing, use size 4 instead of 0 as default packet size

Commit:
Core/Packets: use packet size 4 instead of 0

Shauren commented:
CONNECTION_INSTANCE does not strictly mean inside instance maps but map/world connection

Shauren commented:
Why not automate sending it by placing the call in this method https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Instances/InstanceScript.cpp#L680? You have access to DungeonEncounter dbc entry there

Shauren commented on src/server/game/Server/Packets/InstancePackets.h:
```diff
@@ -236,6 +236,15 @@ namespace WorldPackets
             uint32 CombatResChargeRecovery = 0;
         };
+
+        class InstanceBossKillCredit final : public ServerPacket
```
also please rename this class to BossKillCredit as in SMSG_BOSS_KILL_CREDIT (there is no INSTANCE word here)

Ovahlord commented:
@Shauren I do not automate it because there are many boss encounters that does not use that handling like the Black Temple and some other bosses.

Commit:
Core/Instances: renamed several stuff related to BossKillCredit and use instance connection type instead of realm

Shauren commented:
Maybe its only used for wod+?

Ovahlord commented:
It is used for everything that has been reworked with MoP and also some Cataclysm raid bosses like Al'Akir use that mechanic. Looks to me like Blizz did not finish their job in upgrading all codes because it is all random.

Ovahlord commented:
I would still keep that instance function at place and maybe add another one to send it to single players for world bosses and to send it for non-tracked encounters like in Blackwing Descent.

Commit:
Core/Units: add a function to send boss kill credit to all players within visibility range
*this function can be used for world bosses that are outside of an instanced map

Shauren commented:
While it is useful to have a method allowing to send it for world bosses, its current implementation is incorrect because only players that will receive credit for the kill should get this notification, not every nearby player. Seeing as we currently don't have a way to support that condition it would be best to not add an incorrect implementation - please revert Unit::SendBossKillCredit

Commit:
Revert "Core/Units: add a function to send boss kill credit to all players within visibility range"

This reverts commit cd109b7da698910e1680fc75a2ac61b82a334a0a.
